### PR TITLE
[V1][Spec Decode] EAGLE-3 Support

### PR DESCRIPTION
**Pull request by benchislett**
## Overview

This PR adds EAGLE3 support for llama3 models in the vLLM V1 engine. The main change to existing code is that _auxiliary hidden states must be returned from the target model_. EAGLE3 consumes a concatenation of hidden states extracted from intermediate layers in the target model. This will require modification of any model which wants to support EAGLE3.

## EAGLE3 Model Implementation

The EAGLE reference implementation of EAGLE3 can be found [here](https://github.com/SafeAILab/EAGLE/blob/main/eagle/model/cnets.py). There are a number of changes from previous EAGLE implementations:
- The projection of concatenated [input_embeds, hidden_states] is moved into the attention matrices. Now, the `self.fc` matrix projects the 3x hidden state inputs into a single embedding which is concatenated into [input_embeds, projected_hidden_states]. The final projection to `hidden_dim` happens in the attention as a part of the `QKVParallelLinear` matrices.
- A pair of hidden_states are returned from the EAGLE layer. This is so that the normed value can be used for sampling, and the pre-normalization value can be passed into the next iteration of speculation. An alternative way to handle this is to apply the norm in the compute_logits function. I am indifferent to how we choose to accomplish this.
- Output token mapping must be implemented, as the EAGLE3 lm_head is not the same as the target model and has output dimension 32k which is mapped to the target model vocab size using a simple index mapping (called "d2t" in the EAGLE codebase).

## Acceptance Rate

To generate the acceptance rates, you will need #16367. I merged this at one point for testing, then reverted the diff so it doesn't appear on the PR. For now, you can just check out the previous commit  `git checkout aa11bef1632c345224dda0aa2a56248c5de8ea55`. Alternatively, for latest changes, you can try `git revert 0721cfa` which undoes the revert and returns the experimental diff to the branch.

To run, download the mt_bench data and invoke:
```
VLLM_USE_V1=1 python examples/offline_inference/eagle.py --dataset="./data/mt_bench/question.jsonl" --num_spec_tokens 7 --max_num_seqs 1 --num_prompts 80
```

Here are the results for meta-llama/**Llama-3.1-8B-Instruct** with **yuhuili/EAGLE**-LLaMA3.1-Instruct-8B on my **RTX 4090** (**baseline**):
```
Processed prompts: 100%|...| 80/80 [03:23<00:00,  2.55s/it, est. speed input: 39.49 toks/s, output: 82.47 toks/s]
--------------------------------------------------
mean acceptance length:         2.48
--------------------------------------------------
acceptance at token 0:1.00
acceptance at token 1:0.68
acceptance at token 2:0.39
acceptance at token 3:0.21
acceptance at token 4:0.11
acceptance at token 5:0.06
acceptance at token 6:0.03
acceptance at token 7:0.02
```

Here are the EAGLE3 results with **yuhuili/EAGLE3**-LLaMA3.1-Instruct-8B (**this PR**):
```
Processed prompts: 100%|...| 80/80 [01:56<00:00,  1.46s/it, est. speed input: 68.86 toks/s, output: 143.81 toks/s]
--------------------------------------------------
mean acceptance length:         3.57
--------------------------------------------------
acceptance at token 0:1.00
acceptance at token 1:0.75
acceptance at token 2:0.55
acceptance at token 3:0.42
acceptance at token 4:0.31
acceptance at token 5:0.24
acceptance at token 6:0.18
acceptance at token 7:0.13
```

Notably, the conditional acceptance rate per-token does not seem to decrease. Each token is ~75% likely to be sampled assuming the previous token is accepted. 

Profiling indicates that CUDA graphs and avoiding synchronization in EAGLE code would significantly improve overall performance as drafting time approaches the decode time for increased amounts of speculated tokens. 

## Testing 

I am in the process of updating acceptance tests to evaluate EAGLE3. I hope that code review can be done while test integration is ongoing.

### CODE REVIEW

File: vllm/config.py [hunk 1]
```diff
@@ -2388,9 +2389,10 @@ def __post_init__(self):
                 )
 
                 # Automatically detect the method
-                if self.method == 'eagle':
+                if self.method == 'eagle' or self.method == 'eagle3':
```
Comment (WoosukKwon) [inline vllm/config.py:2392]:
```suggestion
                if self.method in ('eagle', 'eagle3'):
```

File: tests/models/registry.py [hunk 2]
```diff
@@ -391,7 +391,11 @@ def check_available_online(
     "EagleLlamaForCausalLM": _HfExamplesInfo("yuhuili/EAGLE-LLaMA3-Instruct-8B",
                                              trust_remote_code=True,
                                              speculative_model="yuhuili/EAGLE-LLaMA3-Instruct-8B",
-                                             tokenizer="meta-llama/Meta-Llama-3-8B-Instruct"),  # noqa: E501
+                                             tokenizer="meta-llama/Meta-Llama-3-8B-Instruct"),  # noqa: E501,
```
Comment (WoosukKwon) [inline tests/models/registry.py:394]:
```suggestion
                                             tokenizer="meta-llama/Meta-Llama-3-8B-Instruct"),  # noqa: E501
```

File: vllm/model_executor/models/llama.py [hunk 3]
```diff
@@ -113,6 +113,10 @@ def __init__(self,
         super().__init__()
         layer_idx = extract_layer_index(prefix)
         self.hidden_size = hidden_size
+        self.input_hidden_size = hidden_size
```
Comment (benchislett) [inline vllm/model_executor/models/llama.py:116]: nit: I can probably refactor this into `llama_eagle3.py` so there's a smaller diff to the main `llama.py` model file.

Comment (wwl2755) [reply vllm/model_executor/models/llama.py:116]: Agree. I think it would be better to keep llama.py as clean and simple as possible.

File: vllm/config.py [hunk 4]
```diff
@@ -2401,7 +2403,7 @@ def __post_init__(self):
                     self.method = "draft_model"
 
                 # Replace hf_config for EAGLE draft_model
-                if self.method == "eagle":
+                if self.method in ("eagle", "eagle3"):
```
Comment (wwl2755) [inline vllm/config.py:2406]: nit: It seems the current EAGLE3 is only for llama architecture, so maybe we should document this & add an assertion to avoid users from using another model structure.

File: vllm/model_executor/models/llama.py [hunk 5]
```diff
@@ -520,6 +533,13 @@ def __init__(self,
         self.make_empty_intermediate_tensors = (
             self.model.make_empty_intermediate_tensors)
 
+    def set_aux_hidden_state_layers(self, layers: list[int]) -> None:
+        self.model.aux_hidden_state_layers = layers
+
+    def get_eagle3_aux_hidden_state_layers(self) -> list[int]:
+        num_layers = len(self.model.layers)
+        return [2, num_layers // 2, num_layers - 3]
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama.py:541]: Is this same for all sizes of the model?

File: vllm/model_executor/models/llama.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/model_executor/models/llama.py:116]: +1 to @wwl2755. Please move this to the eagle3 file

File: vllm/model_executor/models/llama_eagle3.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+            prefix=maybe_prefix(prefix, "embed_tokens"),
+        )
+        self.config.input_hidden_size = 2 * self.config.hidden_size
+        self.layers = nn.ModuleList([
+            LlamaDecoderLayer(
+                self.config,
+                prefix=maybe_prefix(prefix, f"layers.{start_layer_id}"),
+            )
+        ])
+        if hasattr(self.config, "target_hidden_size"):
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama_eagle3.py:103]: What is this for?

File: vllm/model_executor/models/llama_eagle.py [hunk 7]
```diff
@@ -82,7 +82,7 @@ def forward(
                 hidden_states,
                 residual,
             )
-        return hidden_states + residual
+        return hidden_states + residual, hidden_states + residual
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama_eagle.py:85]:
```suggestion
        hidden_states = hidden_states + residual
        return hidden_states, hidden_states
```

File: vllm/v1/worker/gpu_model_runner.py [hunk 8]
```diff
@@ -161,14 +161,18 @@ def __init__(
 
         # Set up speculative decoding.
         self.use_spec_decode = False
+        self.use_aux_hidden_state_outputs = False
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:164]: nit: Can we rename this?
```suggestion
        self.use_eagle3 = False
```

File: vllm/model_executor/models/llama.py [hunk 9]
```diff
@@ -331,6 +335,8 @@ def __init__(self,
         else:
             self.norm = PPMissingLayer()
 
+        self.aux_hidden_state_layers = []
```
Comment (WoosukKwon) [inline vllm/model_executor/models/llama.py:338]:
```suggestion
        self.aux_hidden_state_layers: list[int] = []
```
Maybe consider using tuple?

File: vllm/model_executor/models/llama.py (hunk 5 above)
Comment (WoosukKwon) [reply vllm/model_executor/models/llama.py:541]: Also, is this same for all models? Or, is it only the case for Llama?

File: vllm/model_executor/models/llama_eagle3.py
Comment (WoosukKwon) [inline vllm/model_executor/models/llama_eagle3.py:1]: Just wondering: What do you think about merging this file with `llama_eagle.py`?

**Review by WoosukKwon — COMMENTED** [review]
@benchislett Thanks for submitting the PR! This is amazing! 🚀

Left some minor comments on the style. Please check out my comments.

**Comment by WoosukKwon** [comment]
@benchislett Also, can you please add correctness tests for EAGLE3?

File: vllm/model_executor/models/llama.py (hunk 5 above)
Comment (benchislett) [reply vllm/model_executor/models/llama.py:541]: This is unclear, as the EAGLE reference code seems to only support Llama models. The relevant LOC is [here](https://github.com/SafeAILab/EAGLE/blob/91ae5f2ffa44a6e2cf50b7a1d19d899c7bbd5817/eagle/model/modeling_llama_kv.py#L1138). It is therefore the same for all Llama models.

I think it's best to keep this as a per-model-class parameter until we have reason to believe otherwise.

File: vllm/v1/worker/gpu_model_runner.py (hunk 8 above)
Comment (benchislett) [reply vllm/v1/worker/gpu_model_runner.py:164]: I think it makes sense to have this functionality distinct from EAGLE3. For now it is the only method that needs hidden state outputs, but in the future there may be other methods which use them in a similar way. I would prefer to keep the `aux_hidden_states_outputs` the same and have eagle-3 be a configuration of that option.

If you feel strongly about this, I can change it.

File: vllm/model_executor/models/llama_eagle3.py
Comment (benchislett) [reply vllm/model_executor/models/llama_eagle3.py:1]: I do not think that would be wise. These are built by extending (overwriting) classes in the Llama code. Eagle3 and Eagle1/2 have very different designs and need to override in different ways, such as EAGLE3 needing to overwrite the input hidden dimension to QKVParallelLinear. It would not be easy to make the implementation so generic that the same class could be used for both, and it is not clean to have separate overrided-implementations in the same file.

File: vllm/model_executor/models/llama.py (hunk 9 above)
Comment (benchislett) [reply vllm/model_executor/models/llama.py:338]: Ah, I thought this was referring to the other variable. Will change to a set.

File: vllm/model_executor/models/llama_eagle3.py
Comment (WoosukKwon) [reply vllm/model_executor/models/llama_eagle3.py:1]: Got it. Thanks for the explanation!

File: vllm/v1/worker/gpu_model_runner.py (hunk 8 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:164]: Ok. I don't have a strong preference either, so let's keep it and refactor it when needed. :)

**Review by WoosukKwon — APPROVED** [review]
@benchislett LGTM. Thanks for the PR! Really excited to have this 🚀
Please merge from main to fix the docker build issue.

**Comment by ekagra-ranjan** [comment]
Thanks for the PR @benchislett . The results you shared in the PR description is using K=7 which would beyond the optimal K and the e2e gain in output token/s as 143/82 = 1.74x. I did a [bench](https://github.com/vllm-project/vllm/pull/16035#issuecomment-2790985075) with K=2 and K=4 and EAGLE 1 had output token/s as ~170 and is faster than EAGLE 3 with K=7. 

Could you pls bench on K=2 and K=4 and share how much faster is EAGLE 3 over EAGLE 1?

**Comment by ekagra-ranjan** [comment]
> but I think the cost should be the same. The acceptance rate should be the only relevant factor
@benchislett - I don't think the cost is the same. The cost to verify tokens would be similar for K=2 or K=7 for lower bs but the the cost to draft tokens will be higher for K=7 compared to K=2 since its an autoregressive process. The earlier bench done on V1 vLLM for Eagle 1 show that K=2 is faster than K=4 end to end even though K=4 has higher AL than K=2.

**Comment by benchislett** [comment]
@ekagra-ranjan I mean to say that the cost of evaluating an EAGLE layer should be the same for EAGLE1 vs EAGLE3. So the performance difference of EAGLE1 for some K and EAGLE3 for the same K will be the relative difference in acceptance lengths for that K. Granted, EAGLE3 might have an optimal perf for a K which is higher than that of EAGLE1.

**Comment by ekagra-ranjan** [comment]
Output tokens/s on H100 BS1

cmd: 
`VLLM_USE_V1=1 python examples/offline_inference/eagle.py --dataset="./data/mt_bench/question.jsonl" --num_spec_tokens 7 --max_num_seqs 1 --num_prompts 80`

vanilla (`meta-llama/Llama-3.1-8B-Instruct`): 133

EAGLE 1 (`yuhuili/EAGLE-LLaMA3.1-Instruct-8B`):
* K=2: 201.69, speedup:  1.51x
* K=4: 198, speedup: 1.48x
* K=7: 169.92, speedup: 1.27x

EAGLE 3 (`yuhuili/EAGLE3-LLaMA3.1-Instruct-8B`):
* K=2: 220.21, speedup: 1.65x, speedup wrt EAGLE 1: 1.09x
* K=4: 228.71, speedup: 1.72x, speedup wrt EAGLE 1: 1.15x
* K=7: 219.38, speedup: 1.65x, speedup wrt EAGLE 1: 1.29x

Trying to measure AL now.

**Comment by ekagra-ranjan** [comment]
I just did the [bench](https://github.com/vllm-project/vllm/pull/16937#issuecomment-2828889840) for K=2,4,7 and compared EAGLE1 and EAGLE3. The AL matches with your run.
* In your RTX 4090 run, EAGLE-3 K=7 gives 1.74x more speedup than EAGLE-1. However, it is 1.29x on H100.
* K=4 was better for E3 and K=2 for E1. The ratio of E3/E1 at their optimal K is 1.14x

@benchislett - what are your thought on the RTX 4090 vs H100 results and this benchmark?

**Comment by DarkLight1337** [comment]
The failing test looks related to this PR, PTAL

**Comment by benchislett** [comment]
@ekagra-ranjan this seems reasonable. Are you measuring output token throughput, or TPOT? These are not quite the same, I believe output token throughput is averaged over the duration of the whole run so prefill time will be included. There will be a slight difference if we measure one way or the other. I re-iterate that the goal of this PR is to establish initial support, and not to optimize the implementation. There is lots of perf to be gained by refining the EAGLE code, but that is outside the scope of this PR.

**Comment by benchislett** [comment]
@DarkLight1337 (cc @WoosukKwon) I'm not sure how to address these failing tests. There are two failures:
- The base EAGLE test is failing with 68% matches compared to the baseline of 70%. I think this might just be variance, as I didn't change much about the EAGLE code and did not significantly change the launch configuration in the test.
- The EAGLE3 launch is hitting OOM. The tests passed on my 4090, so I wonder if this is a matter of the pytorch cleanup not freeing memory between tests or some similar artifact.

Besides re-running or tweaking the acceptance conditions for the tests, I'm not sure how to move forward here

**Comment by ekagra-ranjan** [comment]
>  Are you measuring output token throughput, or TPOT?

@benchislett Thanks for pointing it out. The output tokens/s by running `/offline_inference/eagle.py` is indeed not removing the TTFT. I ran `benchmark_serving.py` to measure the TPOT at BS1 for Llama 3.1 and its 7.5ms which corresponds to 133 tokens/s which is very close to the output tokens/s reported by `/offline_inference/eagle.py`. This is because the input len is small so the numbers I have shared can be treated as TPOT. The TTFT for MTBench is ~12ms so it contributes to <2% even if output len is 100, i.e., [12/(7.5*100)] *100%. 

What are your thoughts on RTX 4090 getting 1.74x gain for E3 over E1 whereas H100 getting only 1.29x? RTX 4090 is getting 1.34x more gains than H100. I tried comparing for the fp16 tensorcore FLOPS/mem BW of RTX 4090 vs H100 hoping that RTX 4090 is relatively more memory bandwidth than H100 but it does seem so. As per [this](https://vast.ai/pricing/gpu/RTX-4090), fp16 tensorcore flops = 330 TFOPS and mem bw is 1 TBPS so the memory bw is just 300x slower than compute flops whereas H100 is 600x slower (1979/3.35).

**Comment by WoosukKwon** [comment]
@benchislett @DarkLight1337 I've merged the PR and submitted a new PR #17209 to fix the test (by just lowering the bar).
I think the failure is because we changed the target model from llama 3 to llama 3.1.

**Comment by linglingfancodeopensource** [comment]
exciting feature
